src/resource.ts

```diff
@@ -1,40 +1,75 @@
 import { createEffect } from './effect';
-import { createSignal, type Signal } from './signal';
-import { createTag, markDependency, markUpdate, type Tag } from './tag';
+import { createSignal } from './signal';
+import { createTag, markDependency, markUpdate } from './tag';
 import type { ReactiveValue } from './types';
 
 type Fetcher<ValueType> = (source: true) => Promise<ValueType>;
 type FetcherWithSource<SourceType, ValueType> = (source: SourceType) => Promise<ValueType>;
 
+// These effectively do what `const enum` does, but doing it manually so we are not coupled to
+// use `tsc` itself to build them. Additionally, though, we only allocate a single array per
+// `Resource`, and reuse its memory when transitioning states. It would be nice if we could
+// make *that* operation safe, but at least this way it is isolated to these "constructors"
+// and the usage can be type safe.
+const LOADING_TAG = 0;
+const LOADED_TAG = 1;
+const ERROR_TAG = 2;
+
+type Loading = [typeof LOADING_TAG];
+const Loading = (): Loading => [LOADING_TAG];
+
+type Loaded<T> = [typeof LOADED_TAG, T];
+const Loaded = <T>(state: State<T>, value: T): Loaded<T> => {
+  state[0] = LOADED_TAG;
+  state[1] = value;
+  return state as Loaded<T>;
+};
+
+type Failed = [typeof ERROR_TAG, unknown];
+const Failed = <T>(state: State<T>, err: unknown): Failed => {
+  state[0] = ERROR_TAG;
+  state[1] = err;
+  return state as Failed;
+};
+
+type State<T> = Loading | Loaded<T> | Failed;
+
 export class Resource<ValueType> {
   private fetcher: Fetcher<ValueType>;
-  private tag: Tag;
-  loading = createSignal(false);
-  error: Signal<any> = createSignal();
+  private tag = createTag();
+  private state = createSignal<State<ValueType>>(Loading());
```

Comment:
This way we only pay for a single `Signal` and a single (reused) `Array`, vs. two `Signal` instances. Objects are *substantially* heavier than `Array`s, so this is a meaningful win, given the in-place rewrite of the state above.